drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,294 @@
+{
+	ARG_UNUSED(arg);
+
+	u32_t cycles;
+	s32_t ticks;
+
+	/*
+	 * Clear RTOS timer interrupt status RW/1C status in GIRQ23 register.
+	 * NVIC will clear its pending bit on ISR exit.
+	 */
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/*
+	 * If we get here and the CCR isn't zero, then this interrupt is
+	 * stale: it was queued while z_clock_set_timeout() was setting
+	 * a new counter. Just ignore it. See above for more info.
+	 */
+	if (RTMR_REGS->CNT != 0) {
+		k_spin_unlock(&lock, key);
+		return;
+	}
+
+	/*
+	 * The original local APIC driver would restart the timer to minimize
+	 * drift. Re-starting here will not work with RTMR.
+	 * Speculation based on experiments with RTMR observing up to one 32KHz
+	 * clock delay in re-starting it.
+	 * 1. RTMR is re-started here.
+	 * 2. This ISR calls z_clock_announce is called.
+	 * 3. z_clock_announce will stop the timer and reprogram it.
+	 * 4. If this happens before 1 tick has elapsed RTMR malfunctions and
+	 *    an extra 32KHz cycle can be introduced.
+	 * Even worse, if z_clock_announce stops the timer by only clearing the
+	 * start bit, the timer will not accept the new preload value and will
+	 * wrap around from zero.
+	 * All we do in the ISR is update total_cycles and last_announcement.
+	 * We depend upon the ISR call to z_clock_announce's call of
+	 * z_clock_set_timeout to restart the timer.
+	 */
+	cycles = cached_icr;
+	cached_icr = MAX_TICKS * CYCLES_PER_TICK;
+	total_cycles += cycles;
+	ticks = (total_cycles - last_announcement) / CYCLES_PER_TICK;
+	last_announcement = total_cycles;
+	k_spin_unlock(&lock, key);
+	z_clock_announce(ticks);
+}
+
+#else
+
+static void isr(void *arg)
+{
+	ARG_UNUSED(arg);
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	total_cycles += CYCLES_PER_TICK;
+	RTMR_REGS->PRLD = cached_icr;
+	RTMR_REGS->CTRL = TIMER_START_VAL;
+	k_spin_unlock(&lock, key);
+
+	z_clock_announce(1);
+}
+
+u32_t z_clock_elapsed(void)
+{
+	return 0U;
+}
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+/*
+ * Return an increasing hardware cycle count.
+ * Implementation: We return current total number of cycles elapsed
+ * from first start of the timer. The return value is 32-bits
+ * resulting in only the lower 32-bits of the total count
+ * being returned.
+ */
+u32_t z_timer_cycle_get_32(void)
+{
+	u32_t ret;
+	u32_t ccr;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	ccr = RTMR_REGS->CNT;
+	ret = total_cycles + (cached_icr - ccr);
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+int z_clock_driver_init(struct device *device)
+{
+	ARG_UNUSED(device);
+
+	mchp_pcr_periph_slp_ctrl(PCR_RTMR, MCHP_PCR_SLEEP_DIS);
+
+	/* Toggle RTMR block enable to clear registers */
+	RTMR_REGS->CTRL = MCHP_RTMR_CTRL_BLK_EN;
+	RTMR_REGS->CNT;
+	RTMR_REGS->CTRL = 0;
+
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
+
+	RTMR_REGS->PRLD = cached_icr;
+
+	IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);
```

Comment:
This is causing a lot of build errors:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/mchp_xec_rtos_timer.c:287:28: error: 'xec_rtos_timer_isr' undeclared (first use in this function)
```

board is mec15xxevb_assy6853

Reply:
@scottwcpg  FYI

TICKLESS_KERNEL(=n) "Tickless kernel"

fails like Andrew said above

[ 34%] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/timer/mchp_xec_rtos_timer.c.obj
/home/francisco/mchip_zephyr/zephyr/drivers/timer/mchp_xec_rtos_timer.c: In function 'z_clock_driver_init':
/home/francisco/mchip_zephyr/zephyr/drivers/timer/mchp_xec_rtos_timer.c:256:151: error: 'xec_rtos_timer_isr' undeclared (first use in this function)
  IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);

Reply:
Done. Checked build with tickless disabled is OK.